File: src/clients/mobile-web/src/features/logs/services/productWaterAffinity.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 *
 * productWaterAffinity — the client mirror of
 * `ShramSafal.Domain.Dfes.ProductWaterAffinity` (founder decision 14, 2026-08-16).
 *
 * WHETHER A PRODUCT OWES THE WATER/CARRIER QUESTION IS DECIDED FROM THE PRODUCT,
 * NEVER FROM `inputs[].method`.
 *
 * The founder's words: "it must understand from the word, or not flagging it anywhere,
 * but keep that fertilizer name — such as a farmer might say '0 52 34 दिल', that means an
 * NPK grade that is given. We already made the AI intelligent enough, don't just confuse
 * it." Follow-up ruling: `fertiliser rule = dry granular`.
 *
 * WHY THIS FILE EXISTS AT ALL. `scoreVlog.ts` runs offline in the browser and cannot call
 * the server, so the rule has to exist on both sides. The two copies are pinned against
 * each other by `productWaterAffinity.parity.test.ts`, which parses the C# source the same
 * way `dfesTuning.test.ts` pins the tuning constants — if either side gains a grade or a
 * granular the other did not, that test fails and names the file.
 *
 * WHY IT IS A NARROW HELPER AND NOT A WIDENING OF `isSprayInput`. `dayState.ts`'s
 * `isSprayInput` feeds `logHasCategoryWork` → `computeScheduleGap`, which is about which
 * SCHEDULED work a log satisfies — a different question with different consequences.
 * Touching it to answer this one would move the schedule engine as a side effect.
 *
 * spec: dfes-companion-2026-07-11 (wave-3.4)
 */

import type { InputEvent } from '../../../types';

export type WaterAffinity = 'WaterCarried' | 'Dry' | 'Unknown';

/**
 * The KNOWN water-soluble NPK grades, in canonical hyphen form.
 * Mirrors `ShramSafal.Domain.Dfes.NpkGradeTable.KnownGrades`.
 *
 * Membership means "a recognised WSF grade". WSFs only exist dissolved, which is what
 * makes the founder's "0 52 34" self-classifying with no flag anywhere.
 */
const KNOWN_GRADES: ReadonlySet<string> = new Set([
    '0-52-34',   // MKP (mono-potassium phosphate)
    '19-19-19',  // balanced NPK WSF
    '0-60-20',   // high-P/K WSF
    '13-0-45',   // KNO3 (potassium nitrate)
    '0-0-50',    // SOP/MOP
    '0-0-60',    // SOP/MOP
]);

/**
 * Recognised grape inputs and their agronomic role.
 * Mirrors `ShramSafal.Domain.Dfes.GrapeProductRoles.Entries` — canonical names only, since
 * the client never sees a raw STT string (the server has already normalised it by the time
 * a log is saved).
 *
 * A role containing "paste" is painted onto the cane and carries no water; everything else
 * here (fungicide / insecticide / PGR / foliar) reaches the vine dissolved.
 */
const GRAPE_PRODUCT_ROLES: ReadonlyMap<string, string> = new Map([
    ['dormex', 'dormancy-break paste'],
    ['ethrel', 'defoliation/ripening'],
    ['6-ba', 'berry sizing'],
    ['cppu', 'berry sizing'],
    ['ga3', 'berry elongation'],
    ['bavistin', 'systemic fungicide'],
    ['curzate', 'downy mildew control'],
    ['alphamethrin', 'pyrethroid insecticide'],
    ['mancozeb', 'contact fungicide'],
    ['copper sulfate', 'Bordeaux mixture input'],
    ['lime', 'Bordeaux mixture alkalizer'],
    ['rally gold', 'systemic fungicide (myclobutanil)'],
    ['pdh', 'potassium di-hydrogen adjuvant'],
]);

/**
 * Broadcast / soil granulars and bulk organics — the founder's `fertiliser rule = dry
 * granular`. Mirrors `ProductWaterAffinity.DryGranulars`.
 *
 * Deliberately SHORT and explicitly named: an unrecognised name must fall through to
 * 'Unknown' and keep asking, never be guessed dry. Every entry added here silences a real
 * question, so it is an agronomic decision and not a convenience.
 */
const DRY_GRANULARS: ReadonlySet<string> = new Set([
    'dap', 'urea', 'mop', 'fym', 'ssp',
    'potash', 'mop potash', 'single super phosphate',
    'di-ammonium phosphate', 'farm yard manure',
    'डीएपी', 'युरिया', 'पोटॅश', 'शेणखत',
]);

/** N-P-K in hyphen or colon form. Anchored: a name that merely CONTAINS three numbers
 *  ("0-52-34 (MKP)", "5:30 वाजता") is not a bare grade and must not resolve at step 1. */
const GRADE_SHAPE = /^\s*\d{1,2}[-:]\d{1,2}[-:]\d{1,2}\s*$/;

function canonicalGrade(s: string | undefined | null): string | undefined {
    if (!s || !GRADE_SHAPE.test(s)) return undefined;
    return s.trim().replace(/:/g, '-');
}

/**
 * Resolve one product row. The resolution order is fixed and total — each step runs only
 * when the one above it did not resolve. See the C# docstring for the full reasoning.
 */
export function resolveWaterAffinity(
    npkGrade: string | undefined | null,
    productName: string | undefined | null,
): WaterAffinity {
    // 1. A recognised NPK grade is water-soluble BY DEFINITION.
    const grade = canonicalGrade(npkGrade) ?? canonicalGrade(productName);
    if (grade !== undefined && KNOWN_GRADES.has(grade)) return 'WaterCarried';

    // 2. A recognised grape input: use its agronomic role.
    const role = productName ? GRAPE_PRODUCT_ROLES.get(productName.trim().toLowerCase()) : undefined;
    if (role !== undefined) {
        return role.toLowerCase().includes('paste') ? 'Dry' : 'WaterCarried';
    }

    // 3. A named dry granular.
    if (productName && DRY_GRANULARS.has(productName.trim().toLowerCase())) return 'Dry';

    // 4. Anything else. Doctrine P4 — do not guess.
    return 'Unknown';
}

/**
 * Does this day owe the water/carrier question?
 *
 * Mirrors `DfesLensExtractor.OwesWater`. The day owes CARRIER when ANY product on it is
 * water-carried, OR when at least one product is UNKNOWN. It owes nothing only when EVERY
 * named product is dry.
 *
 * The asymmetry is deliberate: one unrecognised product is enough to keep the question,
 * because retiring it would remove the farmer's only route to fill that bucket on a day we
 * may simply have failed to recognise (doctrine P4). A day that named no product at all
 * also keeps asking — that is the `length === 0` line, and it is what stops this rule
 * reaching "I sprayed." or a pure-irrigation day.
 */
export function inputsOweWater(inputs: readonly InputEvent[]): boolean {
    const affinities: WaterAffinity[] = [];

    for (const input of inputs) {
        const mix = input.mix ?? [];
        if (mix.length > 0) {
            for (const m of mix) affinities.push(resolveWaterAffinity(m.npkGrade, m.productName));
        } else {
            affinities.push(resolveWaterAffinity(undefined, input.productName));
        }
    }

    if (affinities.length === 0) return true;
    return !affinities.every(a => a === 'Dry');
}

File: src/apps/ShramSafal/ShramSafal.Domain/Dfes/ProductWaterAffinity.cs
// spec: dfes-companion-2026-07-11 (wave-3.4)
using System.Text.RegularExpressions;

namespace ShramSafal.Domain.Dfes;

/// <summary>Whether a product reaches the vine carried by water.</summary>
public enum WaterAffinity
{
    /// <summary>It only reaches the crop dissolved — a spray, a fertigation, a foliar.
    /// The water/carrier question is real and the farmer can answer it.</summary>
    WaterCarried,

    /// <summary>A dry granular broadcast on the soil, or a paste painted on the cane.
    /// There is no tank and no water. The question has no honest answer.</summary>
    Dry,

    /// <summary>We do not recognise this product. <b>Keep asking</b> — see the class
    /// docstring on why this, not "assume dry", is the safe default.</summary>
    Unknown,
}

/// <summary>
/// FOUNDER DECISION 14 (2026-08-16). Whether an input product owes the water/carrier
/// question — decided from the <b>PRODUCT</b>, never from <c>inputs[].method</c>.
///
/// <para><b>The founder's words.</b> "It must understand from the word, or not flagging it
/// anywhere, but keep that fertilizer name — such as a farmer might say '0 52 34 दिल', that
/// means an NPK grade that is given. We already made the AI intelligent enough, don't just
/// confuse it." Follow-up ruling: <c>fertiliser rule = dry granular</c>.</para>
///
/// <para><b>This is NOT the method bypass forbidden on 2026-08-13.</b> That decision
/// answered "should a soil-applied fertiliser still owe the water question?" with "keep
/// asking", and forbade a <c>method == "Soil"</c> test. Decision 14 does not reinstate that
/// test; it removes the method flag from the decision entirely. The difference is visible in
/// behaviour, not just in wording: a 0:52:34 row flagged <c>method: "Soil"</c> still owes
/// water here, and DAP flagged <c>method: "Spray"</c> still does not. A method bypass would
/// get both backwards.</para>
///
/// <para><b>Resolution order is fixed and total</b> — each step is tried only when the one
/// above it did not resolve:</para>
/// <list type="number">
///   <item><b>A recognised NPK grade → WaterCarried.</b> Water-soluble fertilisers only
///   exist dissolved, so a grade that shapes like <c>N-P-K</c> AND appears in
///   <see cref="NpkGradeTable"/> answers the question by itself. This is exactly what makes
///   the founder's "0 52 34" self-classifying with <i>no flag and no AI change</i> — the
///   grade is already emitted verbatim on <c>mix[].npkGrade</c> and persisted to
///   <c>ssf.application_input_items.npk_grade</c>.</item>
///   <item><b>A recognised grape input → its agronomic role.</b> A paste is painted on the
///   cane (Dormex on a dormant vine) and carries no water; a fungicide, insecticide, PGR or
///   foliar reaches the vine dissolved.</item>
///   <item><b>A named dry granular → Dry.</b> The founder's own follow-up rule. A short,
///   explicitly-named list — DAP, urea, MOP, FYM, SSP.</item>
///   <item><b>Anything else → Unknown.</b></item>
/// </list>
///
/// <para><b>Why Unknown keeps asking (doctrine P4).</b> Unknown falls back to today's
/// behaviour: the question stays owed. That will be most logs until the catalogue grows, and
/// it is the correct default in both directions — retiring a question we should have asked
/// silently removes a route for the farmer to raise his number, whereas asking one we could
/// have retired merely leaves an already-existing question in place. Never guess a
/// farmer's day away.</para>
///
/// <para><b><c>method</c> may be a tie-breaker only, never the primary signal.</b> It is not
/// read anywhere in this type. If a later change introduces it, it may only break a tie
/// between two equally-ranked affinities — it may never override one that resolved here,
/// and it may never be consulted before step 1.</para>
/// </summary>
public static partial class ProductWaterAffinity
{
    /// <summary>N-P-K in hyphen or colon form, 1–2 digits per segment and nothing else in
    /// the string. Anchored deliberately: "19-19-19" is a grade, but a product name that
    /// merely CONTAINS three numbers is not.</summary>
    [GeneratedRegex(@"^\s*\d{1,2}[-:]\d{1,2}[-:]\d{1,2}\s*$", RegexOptions.Compiled)]
    private static partial Regex GradeShape();

    /// <summary>
    /// Broadcast / soil granulars and bulk organics — the founder's <c>fertiliser rule =
    /// dry granular</c>. Deliberately a SHORT, explicitly-named list: an unrecognised name
    /// must fall through to <see cref="WaterAffinity.Unknown"/> and keep asking, never be
    /// guessed dry. Every addition here silences a real question, so it is an agronomic
    /// decision rather than a convenience.
    /// </summary>
    private static readonly HashSet<string> DryGranulars =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "DAP", "urea", "MOP", "FYM", "SSP",
            "potash", "MOP potash", "single super phosphate",
            "di-ammonium phosphate", "farm yard manure",
            "डीएपी", "युरिया", "पोटॅश", "शेणखत",
        };

    /// <summary>Resolve one product row. <paramref name="npkGrade"/> is
    /// <c>mix[].npkGrade</c> (emitted verbatim by the extraction contract);
    /// <paramref name="productName"/> is the product as spoken/stored.</summary>
    public static WaterAffinity Resolve(string? npkGrade, string? productName)
    {
        // 1. A recognised NPK grade is water-soluble BY DEFINITION. The grade may arrive
        //    on its own field or as the product name itself ("0-52-34" IS what he called
        //    it), so both are canonicalised and tried.
        var grade = Canonical(npkGrade) ?? Canonical(productName);
        if (grade is not null && NpkGradeTable.KnownGrades.ContainsKey(grade))
            return WaterAffinity.WaterCarried;

        // 2. A recognised grape input: use its agronomic role.
        var role = GrapeProductRoles.Entries
            .FirstOrDefault(e => Matches(e, productName))?.AgronomicRole;
        if (role is not null)
            return role.Contains("paste", StringComparison.OrdinalIgnoreCase)
                ? WaterAffinity.Dry            // painted on the cane, not sprayed
                : WaterAffinity.WaterCarried;  // fungicide / insecticide / PGR / foliar

        // 3. A named dry granular.
        if (!string.IsNullOrWhiteSpace(productName) && DryGranulars.Contains(productName.Trim()))
            return WaterAffinity.Dry;

        // 4. Anything else. P4 — do not guess.
        return WaterAffinity.Unknown;
    }

    /// <summary>Normalises a grade to the hyphen form <see cref="NpkGradeTable"/> is keyed
    /// by, or returns null when the string is not grade-shaped at all. Note this does NOT
    /// strip leading zeros: "00-52-34" is intentionally not "0-52-34" here, because the
    /// STT rescue in Infrastructure already reduces spoken colon triples to the canonical
    /// form before they are persisted.</summary>
    private static string? Canonical(string? s)
        => !string.IsNullOrWhiteSpace(s) && GradeShape().IsMatch(s)
            ? s.Trim().Replace(':', '-')
            : null;

    /// <summary>Exact (case-insensitive) match against the canonical name or any spoken
    /// alias. <b>Deliberately not fuzzy.</b> <c>GrapeInputLexicon</c>'s Levenshtein /
    /// phonetic matcher exists to RESCUE a mangled name into a product; here a near-miss
    /// must fall through to Unknown and keep the question, because a wrong match would
    /// retire a question the farmer could have answered.</summary>
    private static bool Matches(GrapeProduct e, string? productName)
        => !string.IsNullOrWhiteSpace(productName)
           && (string.Equals(e.CanonicalName, productName.Trim(), StringComparison.OrdinalIgnoreCase)
               || e.SttAliases.Contains(productName.Trim(), StringComparer.OrdinalIgnoreCase)
               || e.DevanagariAliases.Contains(productName.Trim(), StringComparer.OrdinalIgnoreCase));
}
